Add slice (#159)
SuperDARNCanada/borealis: docs/source/building_an_experiment.rst

    A dictionary of slice_id : interface_type for each sibling slice in the
    experiment at any given time.

**Not currently supported and will be removed**

wavetype *defaults*
    string for wavetype. The default is SINE. **Not currently supported.**

iwavetable *defaults*
    a list of numeric values to sample from. The default is None. Not currently supported
    but could be set up (with caution) for non-SINE. **Not currently supported.**

qwavetable *defaults*
    a list of numeric values to sample from. The default is None. Not currently supported
    but could be set up (with caution) for non-SINE. **Not currently supported.**

Interfacing Types Between Slices
--------------------------------

The other necessary part of an experiment is specifying how slices will interface with each other. Interfacing in this case refers to how these two components are meant to be run. The following types of interfacing are possible:

1. SCAN.
The scan by scan interfacing allows for slices to run a scan of one slice, followed by a scan of the second. The scan mode of interfacing typically means that the slice will cycle through all of its beams before switching to another slice.

>     There are no requirements for slices interfaced in this manner.

Author: ```suggestion

There are no requirements for slices interfaced in this manner.
```

2. INTTIME.
This type of interfacing allows for one slice to run its integration period (also known as integration time or averaging period), before switching to another slice's integration period. This type of interface effectively creates an interleaving scan where the scans for multiple slices are run 'at the same time', by interleaving the integration times.

>     Slices which are interfaced in this manner must share:

Author: ```suggestion
    
    Slices which are interfaced in this manner must share:
```

Writing an Experiment
---------------------

All experiments must be written as their own class and must be built off of the built-in ExperimentPrototype class.  

This means the ExperimentPrototype class must be imported at the start of the experiment file::

    from experiments.experiment_prototype import ExperimentPrototype

The only experiment-wide attribute that is required to be set by the user
when initializing is the CPID, or control program identifier. This should 
be unique to the experiment. 

Other fields that are possible to be set experiment-wide are:

output_rx_rate 
    The sampling rate of the output data. The default is 10.0e3/3 Hz, or 3.333 kHz.

rx_bandwidth
    The sampling rate of the USRPs (before decimation). The default is 5.0e6 Hz,
    or 5 MHz.

tx_bandwidth
    The output sampling rate of the transmitted signal. The default is 5.0e6 Hz, 
    or 5 MHz.

txctrfreq
    The centre frequency of the transmit chain. The default is 12000.0 kHz, or 

>     12 MHz. Note that this is tuned so will be set to a quantized value, and 

Reviewer: ```suggestion
    12 MHz. Note that this is tuned so will be set to a quantized value, which in general is not exactly 12 MHz, and 
```

    the value can be accessed by the user at this attribute after the experiment 
    begins.

rxctrfreq 
    The centre frequency of the receive chain. The default is 12000.0 kHz, or 

>     12 MHz. Note that this is tuned so will be set to a quantized value, and 

Reviewer: ```suggestion
    12 MHz. Note that this is tuned so will be set to a quantized value, which in general is not exactly 12 MHz, and 
```

Experiment Slices
-----------------

Experiment slices are any individual component of an experiment. An experiment 
slice represents one defined pulse sequence at a single frequency and processed
given certain parameters. A slice is defined using a dictionary and the 
necessary slice keys.

Slice Keys
----------

These are the keys that are set by the user when initializing a slice. Some 
are required, some can be defaulted, and some are set by the experiment 
and are read-only.

**Slice Keys Required by the User**

pulse_sequence *required*
    The pulse sequence timing, given in quantities of tau_spacing, for example
    normalscan = [0, 14, 22, 24, 27, 31, 42, 43]. 

tau_spacing *required*
    multi-pulse increment in us, Defines minimum space between pulses.

pulse_len *required*
    length of pulse in us. Range gate size is also determined by this.

num_ranges *required*
    Number of range gates.

first_range *required*
    first range gate, in km

intt *required or intn required*
    duration of an integration, in ms. (maximum)

intn *required or intt required*

>     number of averages to make a single integration, if intt = None.

Reviewer: ```suggestion
    number of averages to make a single integration, only used if intt = None.
```

beam_angle *required*

>     list of beam directions, in degrees off azimuth. Positive is E of N. Array

Reviewer: This could be misleading. Our main antenna array has 16 elements, but this beam_angle array _also_ has 16 elements, and therefore 16 beams!
```suggestion
    list of beam directions, in degrees off azimuth. Positive is E of N. The beam_angle array
```

    this list gives us the direction of the 0th element in the beam_angle list. It is
    up to the writer to ensure their beam pattern makes sense. Typically beam_order is
    just in order (scanning W to E or E to W, ie. [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10,
    11, 12, 13, 14, 15]. You can list numbers multiple times in the beam_order list,
    for example [0, 1, 1, 2, 1] or use multiple beam numbers in a single
    integration time (example [[0, 1], [3, 4]], which would trigger an imaging
    integration. When we do imaging we will still have to quantize the directions we
    are looking in to certain beam directions.

clrfrqrange *required or txfreq or rxfreq required*
    range for clear frequency search, should be a list of length = 2, [min_freq, max_freq]
    in kHz. **Not currently supported.**

txfreq *required or clrfrqrange or rxfreq required*
    transmit frequency, in kHz. Note if you specify clrfrqrange it won't be used.

rxfreq *required or clrfrqrange or txfreq required*
    receive frequency, in kHz. Note if you specify clrfrqrange or txfreq it won't be used. Only
    necessary to specify if you want a receive-only slice.

**Defaultable Slice Keys**

> acf *defaults*

Reviewer: Does a default of false here make sense? I can point to the epopsound program as experience that it doesn't make sense to default to acfs being off. We  ran for a while before Gareth actually asked, at which point I had to disappoint him, because I had defaulted acfs to not be generated.

Author: I think we could change this eventually if you like, but it will be a more complex default because it can only be True on certain conditions and I think we have yet to set up those conditions (pulse encoding not worked into acfs for example, and I think there is other things to consider). So it would be more confusing to understand, ie. "True on default if x, y , and z otherwise False"

    flag for rawacf and generation. The default is False. If True, the following fields are
    also used:
    - averaging_method (default 'mean')
    - xcf (default True if acf is True)
    - acfint (default True if acf is True)
    - lagtable (default built based on all possible pulse combos)
    - range_sep (will be built by pulse_len to verify any provided value)

acfint *defaults*
    flag for interferometer autocorrelation data. The default is True if acf is True, otherwise
    False.

averaging_method *defaults*

>     a string defining the type of averaging to be done. Current methods are 'mean' or 'median.'

Reviewer: ```suggestion
    a string defining the type of averaging to be done. Current methods are 'mean' or 'median'.
```

    you can offset one slice's sequence from the other by a certain time value so as to not run both
    frequencies in the same pulse, etc. Default is 0 offset.

tx_antennas *defaults*
    The antennas to transmit on, default is all main antennas given max
    number from config.

xcf *defaults*
    flag for cross-correlation data. The default is True if acf is True, otherwise False.

**Read-only Slice Keys**

clrfrqflag *read-only*
    A boolean flag to indicate that a clear frequency search will be done. 
    **Not currently supported.**

cpid *read-only*
    The ID of the experiment, consistent with existing radar control programs.
    This is actually an experiment-wide attribute but is stored within the 
    slice as well. This is provided by the user but not within the slice, 
    instead when the experiment is initialized.

rx_only *read-only*
    A boolean flag to indicate that the slice doesn't transmit, only receives.

slice_id *read-only*
    The ID of this slice object. An experiment can have multiple slices. This 
    is not set by the user but instead set by the experiment when the 

>     slice is added. It will be unique to an experiment run. When experiments

Reviewer: "It will be unique to an experiment run"  This sentence is a bit confusing to me. Could you instead put something like: "Each slice id within an experiment is unique" ?

**Not currently supported and will be removed**

wavetype *defaults*
    string for wavetype. The default is SINE. **Not currently supported.**

iwavetable *defaults*
    a list of numeric values to sample from. The default is None. Not currently supported
    but could be set up (with caution) for non-SINE. **Not currently supported.**

qwavetable *defaults*
    a list of numeric values to sample from. The default is None. Not currently supported
    but could be set up (with caution) for non-SINE. **Not currently supported.**

Interfacing Types Between Slices
--------------------------------

The other necessary part of an experiment is specifying how slices will interface with each other. Interfacing in this case refers to how these two components are meant to be run. The following types of interfacing are possible:

1. SCAN. 
The scan by scan interfacing allows for slices to run a scan of one slice, followed by a scan of the second. The scan mode of interfacing typically means that the slice will cycle through all of its beams before switching to another slice.
    There are no requirements for slices interfaced in this manner.

2. INTTIME. 

> This type of interfacing allows for once slice to run its integration period (also known as integration time or averaging period), before switching to another slice's integration period. This type of interface effectively creates an interleaving scan where the scans for multiple slices are run 'at the same time', by interleaving the integration times.

Reviewer 2: 'for one slice'

    Slices which are interfaced in this manner must share:
        - the same SCANBOUND value.
        - the same INTT or INTN value.
        - the same BEAM_ORDER length (scan length)
        
4. PULSE. 
Pulse interfacing allows for pulse sequences to be run together concurrently. Slices will have their pulse sequences mixed and layered together so that the data transmits at the same time. For example, slices of different frequencies can be mixed simultaneously, and slices of different pulse sequences can also run together at the cost of having more blanked samples. When slices are interfaced in this way the radar is truly transmitting and receiving the slices simultaneously.
    Slices which are interfaced in this manner must share:
        - the same SCANBOUND value.
        - the same INTT or INTN value.
        - the same BEAM_ORDER length (scan length)

>         TODO

Reviewer: Is this TODO complete now, or are you still verifying the requirements for PULSE interfacing?
